File: DSA_Imp/Exercise/ReverseLinkedList.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 * 
 * typedef struct ListNode listnode;
 * 
 * listnode* listnode_new(int val) {
 *     listnode* node = (listnode *) malloc(sizeof(listnode));
 *     node->val = val;
 *     node->next = NULL;
 *     return node;
 * }
 */

/**
 * @input A : Head pointer of linked list 
 * 
 * @Output head pointer of list.
 */
listnode* reverseList(listnode* A) {
    listnode* prev = NULL;  // Initially, the previous node is NULL
    listnode* curr = A;     // Start with the head node
    listnode* next = NULL;  // Temporary pointer to store the next node
    
    // Traverse the list and reverse the links
    while (curr != NULL) {
        next = curr->next;     // Store the next node
        curr->next = prev;     // Reverse the current node's pointer
        prev = curr;           // Move prev forward
        curr = next;           // Move curr forward
    }
    
    // When the loop ends, prev will be pointing to the new head of the reversed list
    return prev;
}

File: LinkedList/SinglyLinkedList/ReverseLinkedList.py
# Input: head = [1,2,3,4,5], left = 2, right = 4
# Output: [1,4,3,2,5]

from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if not head or left == right:
            return head
        
        dummy = ListNode(0)
        dummy.next = head
        prev = dummy
        
        # Move prev to the node just before `left`
        for _ in range(left - 1):
            prev = prev.next
        
        # Start reversing
        current = prev.next
        next_node = None
        
        for _ in range(right - left):
            next_node = current.next
            current.next = next_node.next
            next_node.next = prev.next
            prev.next = next_node
        
        return dummy.next